## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd


df = pd.read_csv('client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
print(df.index)

RangeIndex(start=0, stop=54639, step=1)


In [4]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
df.isnull()

first   last    job  phone  email  client_id  order_id  order_date  \
0      False  False  False  False  False      False     False       False   
1      False  False  False  False  False      False     False       False   
2      False  False  False  False  False      False     False       False   
3      False  False  False  False  False      False     False       False   
4      False  False  False  False  False      False     False       False   
...      ...    ...    ...    ...    ...        ...       ...         ...   
54634  False  False  False  False  False      False     False       False   
54635  False  False  False  False  False      False     False       False   
54636  False  False  False  False  False      False     False       False   
54637  False  False  False  False  False      False     False       False   
54638  False  False  False  False  False      False     False       False   

       order_week  order_year  item_id  category  subcategory  unit_price  \
0           False       False    False     False        False       False   
1           False       False    False     False        False       False   
2           False       False    False     False        False       False   
3           False       False    False     False        False       False   
4           False       False    False     False        False       False   
...           ...         ...      ...       ...          ...         ...   
54634       False       False    False     False        False       False   
54635       False       False    False     False        False       False   
54636       False       False    False     False        False       False   
54637       False       False    False     False        False       False   
54638       False       False    False     False        False       False   

       unit_cost  unit_weight    qty  line_number  
0          False        False  False        False  
1          False        False  False        False  
2          False        False  False        False  
3          False        False  False        False  
4          False        False  False        False  
...          ...          ...    ...          ...  
54634      False        False  False        False  
54635      False        False  False        False  
54636      False        False  False        False  
54637      False        False  False        False  
54638      False        False  False        False  

[54639 rows x 18 columns]

In [7]:
df.shape

(54639, 18)

In [8]:
# What three item categories had the most entries?
# non_null_counts = df.count()
# top_three_categories= non_null_counts.sort_values(ascending=False)
# top_three_categories.head(3)
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [9]:
# For the category with the most entries,
# which subcategory had the most entries?
# subcategory_with_most_entries = df['ca'].value_counts().idxmax()
# subcategory_with_most_entries
df.loc[df['category']== 'consumables','subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [10]:
# Which five clients had the most entries in the data?
# five_clients_with_most_entries = df['client_id'].value_counts()
# five_clients_with_most_entries.name = 'client_id'
# five_clients_with_most_entries.head(5)
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [11]:
# Store the client ids of those top 5 clients in a list.
# five_clients_with_most_entries_to_df = pd.DataFrame(five_clients_with_most_entries)
# five_clients_with_most_entries_to_df.head(5).index.tolist()
list(df['client_id'].value_counts().head(5).index)

[33615, 66037, 46820, 38378, 24741]

In [12]:
# How many total units (the qty column) did the
# client with the most entries order order?
# top_client_id = df['client_id'].value_counts().idxmax()
# top_client_id
# total_units_ordered_by_client_with_most_entries = df[df['client_id'] == top_client_id]['qty'].sum()
# total_units_ordered_by_client_with_most_entries
df.loc[df['client_id']==33615, 'qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [13]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price']*df['qty']
# df.loc[df['unit_price','subcategory', 'line_subtotal' ]].head()
df['line_subtotal'].head()



0    115164.00
1       523.95
2       527.28
3      1056.18
4      3902.00
Name: line_subtotal, dtype: float64

In [14]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping_price(weight):
    if weight > 50:
        # $7 per pound for orders over 50 pounds
        return weight * 7
    else:
        # $10 per pound for items 50 pounds or under
        return weight * 10
df['total_weight'] = df['unit_weight']*df['qty']
df['shipping_price'] = df['total_weight'].apply(calculate_shipping_price)
df['shipping_price'].head()


0    5512.50
1     312.90
2     458.64
3     356.70
4    6500.20
Name: shipping_price, dtype: float64

In [15]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
TAX_RATE = 0.0925
df['total_price'] = df['line_subtotal'] + df['shipping_price'] + (df['line_subtotal'] + df['shipping_price']) * TAX_RATE
df['total_price'].head()


0    131839.076250
1       914.258625
2      1077.117600
3      1543.571400
4     11364.403500
Name: total_price, dtype: float64

In [16]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_price'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  shipping_price    total_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   
3            3        1056.18         35.67          356.70    1543.571400   
4            1        3902.00        928.60         6500.20   11364.403500   

   line_price  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 23 columns]

In [17]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_price']
#df['profit'].sort_values(ascending=False).head()
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price    total_price  line_price  \
0      115164.00        787.50         5512.50  131839.076250    85597.05   
1         523.95         31.29          312.90     914.258625      629.79   
2         527.28         65.52          458.64    1077.117600      765.18   
3        1056.18         35.67          356.70    1543.571400     1077.35   
4        3902.00        928.60         6500.20   11364.403500     8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
# Check your work using the totals above
total_price_4_order_id_2742071 = df.loc[df['order_id'] == 2742071, 'total_price'].sum()
total_price_4_order_id_2742071

152811.897075

In [19]:
total_price_4_order_id_2173913 = df.loc[df['order_id'] == 2173913, 'total_price'].sum()
total_price_4_order_id_2173913

162388.71929999997

In [20]:
total_price_4_order_id_6128929 = df.loc[df['order_id'] == 6128929, 'total_price'].sum()
total_price_4_order_id_6128929

923441.24045

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [21]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
# This is to calculate the quantity purchased by each of the top 5 clients
top_clients_by_quantity = list(df['client_id'].value_counts().head(5).index)
top_clients_by_quantity


# top_clients_data.head(5)


[33615, 66037, 46820, 38378, 24741]

In [22]:
def client_total(client_id, column):
    client_df = df.loc[df['client_id']== client_id, column]
    return client_df.sum()

for client_id in top_clients_by_quantity:
    print(f"{client_id}:${client_total(client_id,'total_price')}")
    

33615:$8377308.572500001
66037:$10259514.801275
46820:$9743794.320474999
38378:$12906550.880874999
24741:$82268891.97985001


In [23]:

# summary_df = pd.DataFrame(top_clients_data)
# # summary_df# This is to filter the original DataFrame to include only transactions from these top clients
# top_clients_data ={'client_id': top_clients_by_quantity}
# columns = ['client_id','qty', 'shipping_price', 'total_price', 'line_price','profit']
# #top_clients_data
# for col in columns:
#     top_clients_data[col] = [client_total(client_id,col) for client_id in top_clients_by_quantity]
# top_clients_data[col]

In [24]:
# def client_data(client_id, column):
#     client_df = df.loc[df['client_id']== client_id, column]
#     return client_df
# # This is to filter the original DataFrame to include only transactions from these top clients
# top_clients_data ={'client_id': top_clients_by_quantity}
# top_clients_data
# # columns = ['client_id','qty', 'shipping_price', 'total_price', 'line_price','profit']
# # # top_clients_data
# # for col in columns:
# #     top_clients_data[col] = [client_data(client_id,col) for client_id in top_clients_by_quantity]
# # top_clients_data[col]

In [31]:
top_clients_data ={'client_id': top_clients_by_quantity}
columns = ['qty', 'shipping_price', 'total_price', 'line_price','profit']

for col in columns:
    top_clients_data[col] = [client_total(client_id,col) for client_id in top_clients_by_quantity]

summary_df = pd.DataFrame(top_clients_data)
summary_df

client_id     qty  shipping_price   total_price   line_price        profit
0      33615   64313      1828984.89  8.377309e+06   6175313.91  2.201995e+06
1      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
2      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
3      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
4      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07

In [32]:
top_clients_data

{'client_id': [33615, 66037, 46820, 38378, 24741],
 'qty': [64313, 43018, 75768, 73667, 239862],
 'shipping_price': [1828984.89,
  1395151.8499999999,
  1601448.8399999999,
  3429455.4000000004,
  5126448.37],
 'total_price': [8377308.572500001,
  10259514.801275,
  9743794.320474999,
  12906550.880874999,
  82268891.97985001],
 'line_price': [6175313.91,
  7004482.98,
  7007191.640000001,
  9634720.979999999,
  45688899.71],
 'profit': [2201994.6625,
  3255031.821275,
  2736602.6804749994,
  3271829.9008750003,
  36579992.26985]}

In [33]:
# This is to format the
# Currency in millions of dollars.
# Function
def to_millions(x):
    return '${:,.2f}M'.format(x / 1000000)
# Application of the function on the total revenue or profit
summary_df['profit'] = pd.to_numeric(summary_df['profit'])
summary_df['profit'] = summary_df['profit'].apply(to_millions).sort_values(ascending=False)
summary_df['profit'].head()

0     $2.20M
1     $3.26M
2     $2.74M
3     $3.27M
4    $36.58M
Name: profit, dtype: object

In [34]:
# This is to rename the columns
# to names suitable for presentation.
summary_df = summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_price': 'Total Price (Millions)',
    'line_price': 'Line Price (Millions)',
    'profit': 'Profit (Millions)'
    
})
# This is to display the summary data frame after renaming columns 
# to names suitable for presentation
# sorted_summary_df=summary_df.sort_values(by='Total Price (Millions)', ascending=False)
summary_df.head()

Client ID     qty  shipping_price  Total Price (Millions)  \
0      33615   64313      1828984.89            8.377309e+06   
1      66037   43018      1395151.85            1.025951e+07   
2      46820   75768      1601448.84            9.743794e+06   
3      38378   73667      3429455.40            1.290655e+07   
4      24741  239862      5126448.37            8.226889e+07   

   Line Price (Millions) Profit (Millions)  
0             6175313.91            $2.20M  
1             7004482.98            $3.26M  
2             7007191.64            $2.74M  
3             9634720.98            $3.27M  
4            45688899.71           $36.58M

In [35]:
# This is to sort the updated data by "Total Profit" from highest to lowest
summary_df.sort_values(by='Profit (Millions)', ascending=False)
summary_df.head()

Client ID     qty  shipping_price  Total Price (Millions)  \
0      33615   64313      1828984.89            8.377309e+06   
1      66037   43018      1395151.85            1.025951e+07   
2      46820   75768      1601448.84            9.743794e+06   
3      38378   73667      3429455.40            1.290655e+07   
4      24741  239862      5126448.37            8.226889e+07   

   Line Price (Millions) Profit (Millions)  
0             6175313.91            $2.20M  
1             7004482.98            $3.26M  
2             7007191.64            $2.74M  
3             9634720.98            $3.27M  
4            45688899.71           $36.58M

In [30]:
# Brief 2-3 sentence summary of my findings.
########################################
# It can be concluded that my top five (5)
# clients data have been consistent all 
# throughout part 4 of this challenge.
# Additionally, it can be said that the 
# use of functions has be found to be
# very useful in this challenge, for witout
#  the creation, definition and use of functions, 
# this challenge would have been way lenghtier 
# and more involved than it has been.
#######################################